compiler/optimizer/Structure.cpp

```diff
@@ -1197,18 +1197,146 @@ static bool findCycle(TR_StructureSubGraphNode *node, TR_BitVector &regionNodes,
    return false;
    }
 
+   /*
+    * findCycle is a depth first traversal of the graph in search for a cycle.
+    * Returns true if a cycle is found, otherwise returns false.
+    */
+static bool findCycle(TR::Compilation *comp, TR_StructureSubGraphNode *origin, TR_BitVector &regionNodes, int32_t entryNode)
+   {
+   TR::StackMemoryRegion memRegion(*(comp->trMemory()));
+
+   int32_t numNodes = comp->getFlowGraph()->getNextNodeNumber();
+
+   /*
+    * First element of the pair is the subgraph node being looked at.
+    * Second element of the pair is a bool that indicates if the node's children have already been processed.
+    */
+   TR::deque<std::pair<TR_StructureSubGraphNode *, bool>, TR::Region&> nodeStack(0, memRegion);
+
+   /*
+    * nodesDiscovered is a bit vector that tracks the path of subgraph nodes from origin to the node being looked at.
+    * If a node is set to Discovered and its next child to examine is also set to Discovered, a cycle has been found.
+    * Processed flag supersedes Discovered flag. If the node is marked as Processed, it counts as being in the Processed state.
+    */
+   TR_BitVector nodesDiscovered(numNodes, memRegion);
+
+   /*
+    * nodesProcessed is a bit vector that tracks which subgraph nodes have been confirmed to not be part of a cycle.
+    * If a node is set to Processed, both that node and its children have already been looked at and a cycle was not found.
+    */
+   TR_BitVector nodesProcessed(numNodes, memRegion);
+
+   /*
+    * Two pairs for each node are added to nodeStack.
+    * The first one pushed has childrenProcessed set to true and the second one has it set to false.
+    * Note that this is a stack so the pair where childrenProcessed is false is pushed second but handled first.
+    * When the node where childrenProcessed is false is being handled, that node's children will be added to nodeStack.
+    * After all the children have been handled and set to Processed the node on the top of the stack will have childrenProcessed
+    * set to true which sets the node to being Processed.
+    */
+   nodeStack.push_front(std::make_pair(origin,true));
+   nodeStack.push_front(std::make_pair(origin,false));
+
+   while (!nodeStack.empty())
+      {
+      TR_StructureSubGraphNode *node = nodeStack.front().first;
```

Comment:
This variable only gets used in the `for` loop below. We can swing down its definition there to avoid needlessly initializing it in case we bail out in one of the `if` blocks below. In general moving the definition of a variable closer to it's first use minimizes the variable scope and makes it easier to reason about the code.

Reply:
`node` is used to get `nodeNum`. `nodeNum` is used in some of those if blocks so it can't be moved any later.

Reply:
You're right! Totally missed that.